tests/tree_inference/test_simulate.py

```diff
@@ -264,3 +265,174 @@ def test_false_homo_mutated(
+    sp_matrix_nx = np.where(sp_matrix_nx == np.inf, -1, sp_matrix_nx)
+    # adjust to SCITE convention of self connected // i.e. only diagonal
+    sp_matrix_nx = np.where(sp_matrix_nx == 0, 1, sp_matrix_nx)
+
+    # run this implementation
+    sp_matrix = sim.floyd_warshall(A)
+
+    assert np.array_equal(sp_matrix, sp_matrix_nx)
+
+
+def test_floyd_warshall_SCITE_example():
+    """Manual test of Floyd Warshall based on example in SCITE pape"""
+    A = np.array([[1, 0, 0, 0], [0, 1, 1, 0], [0, 0, 1, 0], [1, 1, 0, 1]])
+    SP = np.array([[1, -1, -1, -1], [-1, 1, 1, -1], [-1, -1, 1, -1], [1, 1, 2, 1]])
+    sp_matrix = sim.floyd_warshall(A)
+
+    assert np.array_equal(sp_matrix, SP)
+
+
+def test_floyd_warshall_case2():
+    """Manual test of Floyd Warshall"""
+    A = np.array([[1, 1, 1, 0], [0, 1, 0, 1], [0, 0, 1, 0], [0, 0, 1, 1]])
+    SP = np.array([[1, 1, 1, 2], [-1, 1, 2, 1], [-1, -1, 1, -1], [-1, -1, 1, 1]])
+    sp_matrix = sim.floyd_warshall(A)
+
+    assert np.array_equal(sp_matrix, SP)
+
+
+def test_shortest_path_to_ancestry_matrix_SCITE_example():
+    """Manual test of Floyd Warshall"""
+    SP = np.array([[1, -1, -1, -1], [-1, 1, 1, -1], [-1, -1, 1, -1], [1, 1, 2, 1]])
+    ancestor_test_mat = sim.shortest_path_to_ancestry_matrix(SP)
+    ancestor_true_mat = np.array(
+        [[1, 0, 0, 0], [0, 1, 1, 0], [0, 0, 1, 0], [1, 1, 1, 1]]
+    )
+
+    assert np.array_equal(ancestor_test_mat, ancestor_true_mat)
+
+
+def test_built_perfect_mutation_matrix():
+    """Manual test of building mutation matrix from ancestor matrix
+    and cell placement vector sigma by Eqn 11. - SCITE paper example"""
+    ancestor_matrix = np.array([[1, 0, 0, 0], [0, 1, 1, 0], [0, 0, 1, 0], [1, 1, 1, 1]])
+    sigma = np.array([1, 1, 1, 4, 3, 3, 2])
+
+    mutation_matrix = sim.built_perfect_mutation_matrix(4, ancestor_matrix, sigma)
+    mutation_matrix_true = np.array(
+        [
+            [1, 1, 1, 0, 0, 0, 0],
+            [0, 0, 0, 0, 1, 1, 1],
+            [0, 0, 0, 0, 1, 1, 0],
+            [1, 1, 1, 1, 1, 1, 1],
+        ]
+    )
+
+    assert np.array_equal(mutation_matrix, mutation_matrix_true)
+
+
+@pytest.mark.parametrize("seed,", [42])
+@pytest.mark.parametrize("n_nodes,", [3, 10])
+@pytest.mark.parametrize("n_cells", [10, 20])
+@pytest.mark.parametrize(
+    "strategy",
+    [
+        sim.CellAttachmentStrategy.UNIFORM_INCLUDE_ROOT,
+        sim.CellAttachmentStrategy.UNIFORM_EXCLUDE_ROOT,
+    ],
+)
+def test_attach_cells_to_tree_for_strategy_check_bool(
+    n_nodes: int, seed: int, strategy: sim.CellAttachmentStrategy, n_cells: int
+):
+    """Test of cell attachment strategy was respected."""
+
+    rng = random.PRNGKey(seed)
+    tree = random.choice(rng, 2, shape=(n_nodes, n_nodes))
+
+    # nodes need to be their own parent in the SCITE implementation
+    diag_indices = jnp.diag_indices(tree.shape[0])
+    tree = tree.at[diag_indices].set(1)
+
+    mutation_matrix = sim.attach_cells_to_tree(rng, tree, n_cells, strategy)
```

Comment:
The `rng` is used here twice. It should be `split` whenever it's used, so that a fresh one is used for sampling: https://jax.readthedocs.io/en/latest/jax.random.html

Reply:
Oh, yes. - that was unintended. Fixed.